// projen/lib/web/postcss.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.PostCss = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const tailwind_1 = require("./tailwind");
const json_1 = require("../json");
/**
 * Declares a PostCSS dependency with a default config file.
 */
class PostCss {
    constructor(project, options) {
        this.fileName = options?.fileName ?? "postcss.config.json";
        project.addDeps("postcss");
        const config = { plugins: {} };
        if (options?.tailwind ?? true) {
            config.plugins.tailwindcss = {};
            config.plugins.autoprefixer = {};
            this.tailwind = new tailwind_1.TailwindConfig(project, options?.tailwindOptions);
            config.plugins.tailwindcss.config = this.tailwind.fileName;
            project.addDeps("tailwindcss", "autoprefixer");
        }
        this.file = new json_1.JsonFile(project, this.fileName, {
            obj: config,
            marker: false,
        });
        project.npmignore?.exclude(`/${this.fileName}`);
    }
}
_a = JSII_RTTI_SYMBOL_1;
PostCss[_a] = { fqn: "projen.web.PostCss", version: "0.71.82" };
exports.PostCss = PostCss;
