// projen/lib/web/tailwind.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.TailwindConfig = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const json_1 = require("../json");
/**
 * Declares a Tailwind CSS configuration file.
 *
 * There are multiple ways to add Tailwind CSS in your node project - see:
 * https://tailwindcss.com/docs/installation
 *
 * @see PostCss
 */
class TailwindConfig {
    constructor(project, options) {
        this.fileName = options?.fileName ?? "tailwind.config.json";
        this.file = new json_1.JsonFile(project, this.fileName, {
            obj: {
                purge: [],
                darkMode: false,
                theme: {
                    extend: {},
                },
                variants: {
                    extend: {},
                },
                plugins: [],
            },
        });
        project.npmignore?.exclude(`/${this.fileName}`);
    }
}
_a = JSII_RTTI_SYMBOL_1;
TailwindConfig[_a] = { fqn: "projen.web.TailwindConfig", version: "0.71.82" };
exports.TailwindConfig = TailwindConfig;
