// projen/lib/logging.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isEnabled = exports.disable = exports.empty = exports.warn = exports.error = exports.info = exports.verbose = exports.debug = exports.ICON = void 0;
const chalk = require("chalk");
const common_1 = require("./common");
exports.ICON = "👾"; //'✨';
// disable logs if running inside a test
let enabled = common_1.IS_TEST_RUN ? false : true;
function log(isError, color, ...text) {
    // errors are always emitted, even if logs are disabled
    if (!enabled && !isError) {
        return;
    }
    console.error(`${exports.ICON} ${color(...text)}`);
}
function debug(...text) {
    if (process.env.DEBUG) {
        log(false, chalk.gray, ...text);
    }
}
exports.debug = debug;
function verbose(...text) {
    log(false, chalk.white, ...text);
}
exports.verbose = verbose;
function info(...text) {
    log(false, chalk.cyan, ...text);
}
exports.info = info;
function error(...text) {
    log(true, chalk.red, ...text);
}
exports.error = error;
function warn(...text) {
    log(false, chalk.yellow, ...text);
}
exports.warn = warn;
function empty(count = 1) {
    console.log("\n".repeat(count - 1));
}
exports.empty = empty;
function disable() {
    enabled = false;
}
exports.disable = disable;
function isEnabled() {
    return enabled;
}
exports.isEnabled = isEnabled;
