// projen/lib/makefile.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Makefile = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const file_1 = require("./file");
/**
 * Minimal Makefile.
 */
class Makefile extends file_1.FileBase {
    constructor(project, filePath, options = {}) {
        super(project, filePath, options);
        const all = options.all ? options.all : [];
        const rules = options.rules ? options.rules : [];
        rules.forEach((e) => this.validateRule(e));
        this.all = {
            targets: ["all"],
            prerequisites: all,
            phony: true,
        };
        this.rules = [this.all, ...rules];
    }
    /**
     * Add a target to all
     */
    addAll(target) {
        this.all.prerequisites.push(target);
        return this;
    }
    /**
     * Add multiple targets to all
     */
    addAlls(...targets) {
        targets.forEach((e) => this.addAll(e));
        return this;
    }
    /**
     * Add a rule to the Makefile.
     */
    addRule(rule) {
        this.validateRule(rule);
        this.rules.push(rule);
        return this;
    }
    /**
     * Add multiple rules to the Makefile.
     */
    addRules(...rules) {
        rules.forEach((e) => this.addRule(e));
        return this;
    }
    validateRule(rule) {
        if (!rule.targets || !rule.targets.length) {
            throw new Error('"targets" cannot be undefined or empty for items in "rules"');
        }
    }
    synthesizeContent(resolver) {
        const rules = resolver.resolve(this.rules);
        const lines = [
            ...rules.map((rule) => {
                const targets = rule.targets.join(" ");
                const prerequisites = (rule.prerequisites ? rule.prerequisites : []).join(" ");
                const recipe = rule.recipe ? rule.recipe : [];
                const phony = rule.phony ? [`.PHONY: ${targets}`] : [];
                return [
                    ...phony,
                    `${targets}: ${prerequisites}`.trim(),
                    ...recipe.map((step) => `\t${step}`),
                ].join("\n");
            }),
        ];
        return `${lines.join("\n\n")}\n`;
    }
}
_a = JSII_RTTI_SYMBOL_1;
Makefile[_a] = { fqn: "projen.Makefile", version: "0.71.82" };
exports.Makefile = Makefile;
