// projen/lib/project-build.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProjectBuild = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const component_1 = require("./component");
/**
 * Manages a standard build process for all projects.
 *
 * Build spawns these tasks in order:
 * 1. default
 * 2. pre-compile
 * 3. compile
 * 4. post-compile
 * 5. test
 * 6. package
 */
class ProjectBuild extends component_1.Component {
    constructor(project) {
        super(project);
        this.preCompileTask = project.tasks.addTask("pre-compile", {
            description: "Prepare the project for compilation",
        });
        this.compileTask = project.tasks.addTask("compile", {
            description: "Only compile",
        });
        this.postCompileTask = project.tasks.addTask("post-compile", {
            description: "Runs after successful compilation",
        });
        this.testTask = project.tasks.addTask("test", {
            description: "Run tests",
        });
        this.packageTask = project.tasks.addTask("package", {
            description: "Creates the distribution package",
        });
        this.buildTask = project.tasks.addTask("build", {
            description: "Full release build",
        });
        // if this is not subproject, execute the "default" task which will
        // synthesize project files.
        if (!this.project.parent && this.project.defaultTask) {
            this.buildTask.spawn(this.project.defaultTask);
        }
        this.buildTask.spawn(this.preCompileTask);
        this.buildTask.spawn(this.compileTask);
        this.buildTask.spawn(this.postCompileTask);
        this.buildTask.spawn(this.testTask);
        this.buildTask.spawn(this.packageTask);
        // do not allow additional build phases
        this.buildTask.lock();
    }
}
_a = JSII_RTTI_SYMBOL_1;
ProjectBuild[_a] = { fqn: "projen.ProjectBuild", version: "0.71.82" };
exports.ProjectBuild = ProjectBuild;
