// projen/lib/semver.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Semver = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const semver = require("semver");
/**
 * @deprecated This class will be removed in upcoming releases. if you wish to
 * specify semver requirements in `deps`, `devDeps`, etc, specify them like so
 * `express@^2.1`.
 */
class Semver {
    static of(spec) {
        return new Semver(spec);
    }
    /**
     * Latest version.
     */
    static latest() {
        return new Semver("*");
    }
    /**
     * Accept only an exact version
     */
    static pinned(version) {
        return new Semver(version);
    }
    /**
     * Accept any minor version.
     *
     * >= version
     * < next major version
     */
    static caret(version) {
        return new Semver(`^${version}`);
    }
    /**
     * Accept patches.
     *
     * >= version
     * < next minor version
     */
    static tilde(version) {
        return new Semver(`~${version}`);
    }
    constructor(spec) {
        this.spec = spec;
    }
    get version() {
        return semver.minVersion(this.spec)?.version;
    }
}
_a = JSII_RTTI_SYMBOL_1;
Semver[_a] = { fqn: "projen.Semver", version: "0.71.82" };
exports.Semver = Semver;
