// projen/lib/source-code.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.SourceCode = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const component_1 = require("./component");
const textfile_1 = require("./textfile");
/**
 * Represents a source file.
 */
class SourceCode extends component_1.Component {
    constructor(project, filePath, options = {}) {
        super(project);
        this.filePath = filePath;
        this.indentLevel = 0;
        this.indent = options.indent ?? 2;
        this.file = new textfile_1.TextFile(project, filePath, {
            readonly: options.readonly ?? true,
        });
    }
    get marker() {
        return this.file.marker;
    }
    /**
     * Emit a line of code.
     * @param code The contents, if not specified, just adds a newline
     */
    line(code) {
        const spaces = this.indent * this.indentLevel;
        const prefix = " ".repeat(spaces);
        this.file.addLine((prefix + (code ?? "")).trimEnd());
    }
    /**
     * Opens a code block and increases the indentation level.
     *
     * @param code The code before the block starts (e.g. `export class {`).
     */
    open(code) {
        if (code) {
            this.line(code);
        }
        this.indentLevel++;
    }
    /**
     * Decreases the indentation level and closes a code block.
     *
     * @param code The code after the block is closed (e.g. `}`).
     */
    close(code) {
        if (this.indentLevel === 0) {
            throw new Error("Cannot decrease indent level below zero");
        }
        this.indentLevel--;
        if (code) {
            this.line(code);
        }
    }
}
_a = JSII_RTTI_SYMBOL_1;
SourceCode[_a] = { fqn: "projen.SourceCode", version: "0.71.82" };
exports.SourceCode = SourceCode;
