// projen/lib/cdk8s/cdk8s-app-py.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Cdk8sPythonApp = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdk8s_deps_py_1 = require("./cdk8s-deps-py");
const __1 = require("..");
const python_project_1 = require("../python/python-project");
/**
 * CDK8s app in Python
 *
 *
 * @pjid cdk8s-app-py
 */
class Cdk8sPythonApp extends python_project_1.PythonProject {
    constructor(options) {
        super({ ...options, pytest: false, sample: false });
        this.cdk8sDeps = new cdk8s_deps_py_1.Cdk8sDepsPy(this, {
            dependencyType: __1.DependencyType.RUNTIME,
            cdk8sCliDependency: false,
            ...options,
        });
        if (!options.cdk8sVersion) {
            throw new Error("Required field cdk8sVersion is not specified.");
        }
        this.appEntrypoint = options.appEntrypoint ?? "app.py";
        const synth = this.addTask("synth", {
            description: "Synthesizes your cdk8s app into dist",
            exec: "cdk8s synth",
        });
        // add synth to the build
        this.postCompileTask.spawn(synth);
        const cdk8sImports = options.cdk8sImports ?? [];
        const k8sSpec = options.k8sSpecVersion
            ? `k8s@${options.k8sSpecVersion}`
            : "k8s";
        new __1.YamlFile(this, "cdk8s.yaml", {
            committed: true,
            editGitignore: true,
            obj: {
                language: "python",
                app: `python ${this.appEntrypoint}`,
                imports: [k8sSpec, ...cdk8sImports],
            },
        });
        if (options.sample ?? true) {
            new AppCode(this, this.appEntrypoint);
        }
    }
}
_a = JSII_RTTI_SYMBOL_1;
Cdk8sPythonApp[_a] = { fqn: "projen.cdk8s.Cdk8sPythonApp", version: "0.71.82" };
exports.Cdk8sPythonApp = Cdk8sPythonApp;
class AppCode extends __1.Component {
    constructor(project, filename) {
        super(project);
        new __1.SampleFile(project, filename, {
            contents: [
                "from constructs import Construct",
                "from cdk8s import App, Chart",
                "",
                "",
                "class MyChart(Chart):",
                "  def __init__(self, scope: Construct, id:str):",
                "    super().__init__(scope, id)",
                "",
                "",
                "app = App()",
                'MyChart(app, "${project.name}")',
                "",
                "app.synth()",
            ].join("\n"),
        });
    }
}
