// projen/lib/javascript/npm-config.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.NpmConfig = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const component_1 = require("../component");
const ini_1 = require("../ini");
/**
 * File representing the local NPM config in .npmrc
 */
class NpmConfig extends component_1.Component {
    constructor(project, options = {}) {
        super(project);
        /**
         * The config object. This object can be mutated until the project is
         * synthesized.
         */
        this.config = {};
        new ini_1.IniFile(project, ".npmrc", {
            obj: this.config,
            omitEmpty: options.omitEmpty,
        });
        if (options.registry) {
            this.addRegistry(options.registry);
        }
    }
    /**
     * configure a scoped registry
     *
     * @param url the URL of the registry to use
     * @param scope the scope the registry is used for; leave empty for the default registry
     */
    addRegistry(url, scope) {
        this.addConfig(scope ? `${scope}:registry` : "registry", url);
    }
    /**
     * configure a generic property
     *
     * @param name the name of the property
     * @param value the value of the property
     */
    addConfig(name, value) {
        this.config[name] = value;
    }
}
_a = JSII_RTTI_SYMBOL_1;
NpmConfig[_a] = { fqn: "projen.javascript.NpmConfig", version: "0.71.82" };
exports.NpmConfig = NpmConfig;
