// projen/lib/python/setuptools.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Setuptools = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const setuppy_1 = require("./setuppy");
const component_1 = require("../component");
const dependencies_1 = require("../dependencies");
/**
 * Manages packaging through setuptools with a setup.py script.
 */
class Setuptools extends component_1.Component {
    constructor(project, 
    //moduleName: string,
    options) {
        super(project);
        project.deps.addDependency("wheel@0.36.2", dependencies_1.DependencyType.DEVENV);
        project.deps.addDependency("twine@3.3.0", dependencies_1.DependencyType.DEVENV);
        project.packageTask.exec("python setup.py sdist bdist_wheel");
        this.publishTestTask = project.addTask("publish:test", {
            description: "Uploads the package against a test PyPI endpoint.",
            exec: "twine upload --repository-url https://test.pypi.org/legacy/ dist/*",
        });
        this.publishTask = project.addTask("publish", {
            description: "Uploads the package against a test PyPI endpoint.",
            exec: "twine upload dist/*",
        });
        const packages = options.packageName ? [options.packageName] : undefined;
        new setuppy_1.SetupPy(project, {
            name: project.name,
            packages: packages,
            authorName: options.authorName,
            authorEmail: options.authorEmail,
            version: options.version,
            description: options.description,
            license: options.license,
            homepage: options.homepage,
            classifiers: options.classifiers,
            ...options.setupConfig,
        });
    }
}
_a = JSII_RTTI_SYMBOL_1;
Setuptools[_a] = { fqn: "projen.python.Setuptools", version: "0.71.82" };
exports.Setuptools = Setuptools;
