// projen/lib/vscode/extensions.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.VsCodeRecommendedExtensions = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const component_1 = require("../component");
const json_1 = require("../json");
/**
 * VS Code Workspace recommended extensions
 * Source: https://code.visualstudio.com/docs/editor/extension-marketplace#_workspace-recommended-extensions
 */
class VsCodeRecommendedExtensions extends component_1.Component {
    constructor(vscode) {
        super(vscode.project);
        /**
         * List of extensions which should be recommended for users of this workspace.
         * The identifier of an extension is always '${publisher}.${name}'. For example: 'ms-python.python'.
         *
         * @default - no recommendations are made
         */
        this.recommendations = [];
        /**
         * List of extensions recommended by VS Code that should not be recommended for users of this workspace.
         * The identifier of an extension is always '${publisher}.${name}'. For example: 'ms-python.python'.
         *
         * @default - no recommendations are marked unwanted
         */
        this.unwantedRecommendations = [];
        new json_1.JsonFile(vscode.project, ".vscode/extensions.json", {
            omitEmpty: true,
            allowComments: true,
            obj: {
                recommendations: this.recommendations,
                unwantedRecommendations: this.unwantedRecommendations,
            },
        });
    }
    /**
     * Adds a list of VS Code extensions as recommendations for this workspace.
     *
     * @param extensions The extension IDs
     */
    addRecommendations(...extensions) {
        this.recommendations.push(...extensions);
    }
    /**
     * Marks a list of VS Code extensions as unwanted recommendations for this workspace.
     * VS Code should not be recommend these extensions for users of this workspace.
     *
     * @param extensions The extension IDs
     */
    addUnwantedRecommendations(...extensions) {
        this.unwantedRecommendations.push(...extensions);
    }
}
_a = JSII_RTTI_SYMBOL_1;
VsCodeRecommendedExtensions[_a] = { fqn: "projen.vscode.VsCodeRecommendedExtensions", version: "0.71.82" };
exports.VsCodeRecommendedExtensions = VsCodeRecommendedExtensions;
